cibuildwheel/pyodide.py

```diff
@@ -0,0 +1,342 @@
+from __future__ import annotations
+
+import contextlib
+import os
+import shutil
+import sys
+from collections.abc import Sequence, Set
+from dataclasses import dataclass
+from pathlib import Path
+
+from filelock import FileLock
+
+from .architecture import Architecture
+from .environment import ParsedEnvironment
+from .logger import log
+from .options import Options
+from .typing import PathOrStr
+from .util import (
+    CIBW_CACHE_PATH,
+    BuildSelector,
+    NonPlatformWheelError,
+    call,
+    download,
+    extract_zip,
+    find_compatible_wheel,
+    get_build_verbosity_extra_flags,
+    get_pip_version,
+    prepare_command,
+    read_python_configs,
+    shell,
+    split_config_settings,
+    test_fail_cwd_file,
+    virtualenv,
+)
+
+
+@dataclass(frozen=True)
+class PythonConfiguration:
+    version: str
+    identifier: str
+    pyodide_version: str
+    emscripten_version: str
+
+
+def install_emscripten(tmp: Path, version: str) -> Path:
+    url = "https://github.com/emscripten-core/emsdk/archive/main.zip"
+    installation_path = CIBW_CACHE_PATH / ("emsdk-" + version)
+    emsdk_path = installation_path / "emsdk-main/emsdk"
+    emcc_path = installation_path / "emsdk-main/upstream/emscripten/emcc"
+    with FileLock(str(installation_path) + ".lock"):
+        if installation_path.exists():
+            return emcc_path
+        emsdk_zip = tmp / "emsdk.zip"
+        download(url, emsdk_zip)
+        installation_path.mkdir()
+        extract_zip(emsdk_zip, installation_path)
+        os.chmod(emsdk_path, 0o777)
+        call(emsdk_path, "install", version)
+        call(emsdk_path, "activate", version)
+
+    return emcc_path
+
+
+def install_xbuildenv(env: dict[str, str], pyodide_version: str) -> None:
+    xbuildenv_cache_dir = CIBW_CACHE_PATH / f"pyodide-xbuildenv-{pyodide_version}"
+    pyodide_root = xbuildenv_cache_dir / ".pyodide-xbuildenv/xbuildenv/pyodide-root/"
+    if pyodide_root.exists():
+        env["PYODIDE_ROOT"] = str(pyodide_root)
+        return
+
+    xbuildenv_cache_dir.mkdir(exist_ok=True)
+    env.pop("PYODIDE_ROOT", None)
+    call(
+        "pyodide",
+        "xbuildenv",
+        "install",
+        "--download",
+        env=env,
+        cwd=xbuildenv_cache_dir,
+    )
+    env["PYODIDE_ROOT"] = str(pyodide_root)
+
+
+def get_base_python(identifier: str) -> Path:
+    implementation_id = identifier.split("-")[0]
+    majorminor = implementation_id[len("cp") :]
+    major_minor = f"{majorminor[0]}.{majorminor[1:]}"
+    python_name = f"python{major_minor}"
+    which_python = shutil.which(python_name)
+    if which_python is None:
+        print(
+            f"Error: CPython {major_minor} is not installed.",
+            file=sys.stderr,
+        )
+        raise SystemExit(1)
+    return Path(which_python)
```

Comment:
I would prefer to leave it as is for this PR and if you like we can switch to a different approach in a followup. I think it would add unneeded complexity to this PR. Plus a lot of common CI setups have Python 3.11 preinstalled.

Reply:
I'm in favor of this method, at least until something like PyBI comes along. Saving time is worthwhile, I think; at least for now, requiring Python 3.11 isn't that bad for most CIs, and this is already something you have to setup manually.